contracts/captcha/src/lib.rs

```diff
@@ -223,7 +220,7 @@ pub mod prosopo {
     #[derive(PartialEq, Debug, Eq, Clone, Copy, scale::Encode, scale::Decode)]
     #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
     pub struct LastCorrectCaptcha {
-        pub before_ms: u64,
+        pub before: BlockNumber,
```

Comment:
I think timestamp might be better for this so that we can ask for last captcha within `x` seconds.

Reply:
This code is why we had `before_ms`:

https://github.com/prosopo/dapp-example/blob/08c56990cd6f50ac7068d9351e1792e02ee0d591/contracts/lib.rs#LL104C1-L111C10

Now that it's blocks, any caller of this function will have to specify a recency in blocks. Is it better for a user of this data to specify recency in blocks or as a time interval? The latter would mean doing any kind of time calculations off-chain or estimating them inside the calling contract using a hard coded estimated block time.

Reply:
This got changed due to a knock on effect, which I'll explain first:

I had to change the user_history_max_age to be specified as a # of blocks rather than in milliseconds. This is because the contract does not know how long a block is, so when comparing a block range to milliseconds the comparison did not work. I made the decision to use blocks going forward instead of milliseconds.

This then meant that the LastCorrectCaptcha could not be computed in milliseconds as we only know the block difference, not the ms difference, hence the change.

I'm not opposed to one way or the other, I just defaulted to blocks because it fit the user history limits nicely but had this knock on effect which I did not foresee. We can still use milliseconds, but I do not know how we would do this without knowing the block timestamps to do checks, etc.

For instance, when a provider submits a commit we need to check the recency using block numbers. The provider could also supply a timestamp for the commit in ms, but we would need to check this landed in the block they're claiming the commit is from, which means we need to know the block time again. We can put the block time in the contract and use that?

Perhaps discuss on slack this a.m?

Reply:
After discussion we're proceeding with using # of blocks only, can do conversion to ms externally to the contract and will look to do that inside the contract in future